src/basic/cgroup-util.c

```diff
@@ -2385,12 +2385,18 @@ int cg_enable_everywhere(CGroupMask supported, CGroupMask mask, const char *p) {
         return 0;
 }
 
+static thread_local bool force_legacy = false;
+
 bool cg_is_unified_wanted(void) {
         static thread_local int wanted = -1;
         int r;
         bool b;
         const bool is_default = DEFAULT_HIERARCHY == CGROUP_UNIFIED_ALL;
 
+        /* If this is set, it overrides any detection logic */
+        if (force_legacy)
+                return false;
```

Comment:
is this really necessary? I mean, the mount table is actually quite powerful already: we could simply drop the MNT_FATAL flag from the cgroupsv2 mount. That would mean we'd simply skip to the next item in the table if we were asked to mount unified but couldn't. The nex titem happens to be the tmpfs mount on /sys/fs/cgroup, which we should simply drop the conditionalization for. Since mount table items that are already mount points are silently skipped this would then mean it would only have an effect if the cgroupsv2 item before either wasn't requested or didn't work. Finally, the cg_is_xyz_wanted() calls should be fixed to always check what /sys/fs/cgroup is mounted as and limit their results depending on what they see.

With that in place the logic would be encapsulated entirely in the existing mount setup logic, and I think we'd get a much nicer and safer logic into place...

I hope you can follow what I am babbling about here...